lib/quirc.c

```diff
@@ -44,11 +48,40 @@ void quirc_destroy(struct quirc *q)
 	free(q);
 }
 
+int quirc_init(struct quirc *q, int w, int h, uint8_t *image)
+{
+	if (!q || !image || !QUIRC_PIXEL_ALIAS_IMAGE)
+		return -1;
+
+	/*
+	 * Same sanity check as in quirc_resize() for the same reasons.
+	 */
+	if (w < 0 || h < 0)
+		return -1;
+
+	memset(q, 0, sizeof(*q));
+	q->w = w;
+	q->h = h;
+	q->image = image;
+	return 0;
+}
+
 int quirc_resize(struct quirc *q, int w, int h)
 {
 	uint8_t		*image  = NULL;
 	quirc_pixel_t	*pixels = NULL;
 
+	if (!q->need_to_free) {
+		/*
+		 * Just update the internal sizes and assume the caller knows what they
+		 * are doing and provided a big enough image buffer in quirc_init().
```

Comment:
This is making me nervous. I'm not sure I see the use case for calling `quirc_resize` with a static buffer, but my guess would be that only shrinking should be allowed. In other words, I would check that `w <= q->w && h <= q->h` hold.

Reply:
Imagine a device with a front and rear camera with different resolutions.  You'll only use one at a time, so you could use the same memory buffer address for both (say it's big enough for the higher-resolution camera).  You can then use `quirc_resize()` to switch back and forth between them.

I'm not using this functionality so it's not important to me ATM, but IIRC from previous comments you wanted the API use to be as similar as possible between the two allocation strategies.